File: 290. Word Pattern/Solution.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        char[] letters = pattern.toCharArray();
        String[] words = s.split(" ");
        Set<Character> uniqueLetters = new HashSet<>();
        Set<String> uniqueWords = new HashSet<>();
        for (char l : letters) uniqueLetters.add(l);
        for (String w : words) uniqueWords.add(w);
        if (letters.length != words.length || uniqueLetters.size() != uniqueWords.size()) return false;
        int n = letters.length;
        Map<Character, String> seen = new HashMap<>();
        for (int i = 0; i < n; i++) {
            if (seen.containsKey(letters[i])) {
                if (!seen.get(letters[i]).equals(words[i])) return false;
            } else {
                seen.put(letters[i], words[i]);
            }
        }
        return true;
    }
}

File: 290. Word Pattern/Solution.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words, letters = s.split(" "), [*pattern]
        if len(letters) != len(words) or len(set(letters)) != len(set(words)): return False
        seen = dict()
        n = len(letters)
        for i in range(n):
            if letters[i] in seen:
                if seen[letters[i]] != words[i]: return False
            else:
                seen[letters[i]] = words[i]
        return True
